hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/record/GenesisRecordsConsensusHookImpl.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.workflows.handle.record;
+
+import static com.hedera.node.app.spi.HapiUtils.FUNDING_ACCOUNT_EXPIRY;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.Duration;
+import com.hedera.hapi.node.base.Transaction;
+import com.hedera.hapi.node.state.token.Account;
+import com.hedera.hapi.node.token.CryptoCreateTransactionBody;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.token.records.GenesisAccountRecordBuilder;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.app.spi.workflows.record.GenesisRecordsConsensusHook;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This class is responsible for storing the system accounts created during node startup, and then creating
+ * the corresponding synthetic records when a consensus time becomes available.
+ */
+public class GenesisRecordsConsensusHookImpl implements GenesisRecordsConsensusHook {
+    private static final Logger log = LogManager.getLogger(GenesisRecordsConsensusHookImpl.class);
+    private static final String SYSTEM_ACCOUNT_CREATION_MEMO = "Synthetic system creation";
+    private static final String STAKING_MEMO = "Release 0.24.1 migration record";
+    private static final String TREASURY_CLONE_MEMO = "Synthetic zero-balance treasury clone";
+
+    private Map<Account, CryptoCreateTransactionBody.Builder> systemAccounts = new HashMap<>();
+    private Map<Account, CryptoCreateTransactionBody.Builder> stakingAccounts = new HashMap<>();
+    private Map<Account, CryptoCreateTransactionBody.Builder> miscAccounts = new HashMap<>();
+    private Map<Account, CryptoCreateTransactionBody.Builder> treasuryClones = new HashMap<>();
```

Comment:
nit : we probably don't need to initialize these and have as null if not set. Then we can just check if they are not null in process